Promote "Metadata in Table Schema" recipe to the specs.

pierrecamilleri opened the pull request:
Context and related issue : 
fix [#899](https://github.com/frictionlessdata/specs/issues/899#issuecomment-2076509819)

Migrated [pull request from datapackage-v2-draft repo](https://github.com/frictionlessdata/datapackage-v2-draft/pull/63)

Still to do on this PR:

- [x] Move table schema to /standards/table-schema.md (cf. main)
- [X] Move changelog to /overview/changelog.md (cf. main)
- [x] Update wording so it aligns with Data Package (my [pending PR](https://github.com/peterdesmet/datapackage-1/blob/wording/content/docs/specifications/table-schema.md))
- [x] Update some links
- [x] Fix invalid json example (it was an unintended change)
- [x] Clarify if examples are borrowing title and path from Data Resource (I think they are, included in my PR) or if this is something completely separate.

pierrecamilleri committed:
Schema Metadata spec proposal

pierrecamilleri committed:
🔵 wording and typos

pierrecamilleri committed:
taking into account review comments

pierrecamilleri committed:
Remove `path` property

pierrecamilleri committed:
Remove `updated` property

pierrecamilleri committed:
inherits "if distribted in a data package descriptor"

pierrecamilleri committed:
resource -> example resource

pierrecamilleri committed:
`name` "as for Data Package

pierrecamilleri committed:
Remove sources and build profile

pierrecamilleri committed:
`examples` are `Data Resource`s

pierrecamilleri committed:
Attempt at removing "schema" property from table schema example
resources

pierrecamilleri committed:
Revert "Attempt at removing "schema" property from table schema example"

This reverts commit cc10899d176132160da95ef285b4639b2d165445.

pierrecamilleri committed:
Revert "`examples` are `Data Resource`s"

This reverts commit 8cfb65402be4b8fb29f1a2094d8ec7783b38e1f2.

pierrecamilleri committed:
Clear-up examples documentation (ref to data files)

pierrecamilleri committed:
fix: created stores only creation date

pierrecamilleri committed:
fix: link

pierrecamilleri committed:
changelog

pierrecamilleri committed:
fix: unintended change

pierrecamilleri committed:
fix: link

roll commented:
Thanks a lot for rebasing the PR @pierrecamilleri !

pierrecamilleri committed:
Update doc with Peters suggestions + fix links

pierrecamilleri committed:
"partial Data Resource" ?

pierrecamilleri commented:
@peterdesmet @roll 

Almost there.

I [added "partial"](https://github.com/frictionlessdata/datapackage/pull/961/commits/800593f80045fb31b978381f8fdb3aad0541beaa) to qualify "Data Resource" in  the suggestion of Peter concerning the documentation of top-level "examples" ; because an object with `title` and `path` is not a valid Data Resource (`name` is missing). Or do we want to impose a `name` property ? What do you think of this wording ?

peterdesmet committed:
Use title case for Data Package

peterdesmet committed:
Correct typo + align Data Package

peterdesmet committed:
Require name for examples

peterdesmet commented on profiles/source/dictionary/common.yaml:
```diff
@@ -72,7 +72,7 @@ example:
 homepage:
   title: Home Page
-  description: The home on the web that is related to this data package.
+  description: The home on the web that is related to this resource.
```
In `keywords` the common term `descriptor` is used. That is probably preferred over `resource`, since resource is often used for Data Resource.
```suggestion
  description: The home on the web that is related to this descriptor.
```

peterdesmet commented on profiles/source/dictionary/schema.yaml:
```diff
@@ -263,6 +281,33 @@ tableSchemaMissingValues:
       }
+tableSchemaExamples:
+  title: Examples
+  description: Links to example data files
```
```suggestion
  description: A list of Data Resources that use and illustrate the schema.
```

peterdesmet commented on profiles/source/dictionary/schema.yaml:
```diff
@@ -263,6 +281,33 @@ tableSchemaMissingValues:
+      {
+        "examples": [
+          {
+            "title": "Valid data",
```
```suggestion
            "name": "valid-data",
```

peterdesmet commented on profiles/source/dictionary/schema.yaml:
```diff
@@ -263,6 +281,33 @@ tableSchemaMissingValues:
+      }
+tableSchemaExample:
+  title: Example
+  description: Link to an example data file
```
```suggestion
  description: A Data Resource that uses and illustrates the schema.
```

peterdesmet commented on profiles/source/dictionary/schema.yaml:
```diff
@@ -263,6 +281,33 @@ tableSchemaMissingValues:
+  type: object
+  properties:
+    title:
+      "$ref": "#/definitions/title"
```
```suggestion
      "$ref": "#/definitions/name"
```

pierrecamilleri replied:
Would you mind keeping both `title` and `name` ? Both are present in data resources (`name` is mandatory), but as mentioned in another comment, we use `title` to give hints about what is interesting about the given example. This is much less easy with only the `title`.

EDIT : after writing this, I realized that we can optionally enforce this via a custom profile, so OK for keeping only `name` and `path`.

peterdesmet commented on profiles/source/dictionary/schema.yaml:
```diff
@@ -263,6 +281,33 @@ tableSchemaMissingValues:
+      "$ref": "#/definitions/title"
+    path:
+      "$ref": "#/definitions/path"
+  required: ["title", "path"]
```
```suggestion
  required: ["name", "path"]
```

peterdesmet commented on profiles/target/2.0/datapackage.json:
```diff
@@ -57,7 +57,7 @@
       "propertyOrder": 60,
       "title": "Home Page",
-      "description": "The home on the web that is related to this data package.",
+      "description": "The home on the web that is related to this resource.",
```
Two notes:

1. These files should be rebuild based on the changes above.
2. Is it a good idea to rebuild these files? I think we should have a new directory `target/2.1-draft`

pierrecamilleri replied:
I am not familiar with the build process, in particular what is the difference between "build/profiles" and "profiles/target" ?

When I run the "build.js", with "VERSION=2.1-draft", it does not create (or for the matter, populate) a `target/2.1-draft` directory :thinking:

roll replied:
@peterdesmet 
@pierrecamilleri 
I need to set up the branching process for new versions so it's OK if this PR just includes actual changes to the standard and I'll handle the rest

peterdesmet reviewed (changes requested):
@pierrecamilleri thanks for the updates! Regarding:

> Or do we want to impose a name property ? What do you think of this wording ?

I think we should simply stick to the Data Resource spec and require a `name` (over a `title`). I have made suggestions to the files to reflect this change.

pierrecamilleri committed:
Update profiles/source/dictionary/common.yaml

Co-authored-by: Peter Desmet <peter.desmet.work@gmail.com>

pierrecamilleri committed:
Update profiles/source/dictionary/schema.yaml

Co-authored-by: Peter Desmet <peter.desmet.work@gmail.com>

pierrecamilleri committed:
Update profiles/source/dictionary/schema.yaml

Co-authored-by: Peter Desmet <peter.desmet.work@gmail.com>

pierrecamilleri committed:
Update profiles/source/dictionary/schema.yaml

Co-authored-by: Peter Desmet <peter.desmet.work@gmail.com>

pierrecamilleri committed:
Update : example title to example name

pierrecamilleri commented:
Little ping @peterdesmet, if you can find the time for a review it would be great !

peterdesmet commented:
Thanks for the heads up! Will have a look this week.

roll commented:
Amazing @pierrecamilleri!

I'm going to be on it next week as well. First of all, I would suggest us to test a profile patch on https://github.com/frictionlessdata/datapackage/pull/975, and then do this one

peterdesmet committed:
Rephrase #961 changes

peterdesmet commented on content/docs/standard/table-schema.md:
```diff
@@ -266,6 +266,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
 
+#### `name`
+
+A simple name or identifier for the schema (cf. [Data Package](https://datapackage.org/standard/data-package/#name)).
```
```suggestion
A simple name or identifier for the schema.
```

pierrecamilleri replied:
I am closing this thread and the following ones so that all the threads can be resolved. In the absence of a reply to [my message](https://github.com/frictionlessdata/datapackage/pull/961#issuecomment-2374448138) below, I have decided to keep the text as it is. In any case, the wording can still evolve, including in a subsequent PR, without changing the nature of this spec, so we can still discuss it.

peterdesmet commented on content/docs/standard/table-schema.md:
```diff
@@ -266,6 +266,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `description`
+       
+A description of the schema (cf. [Data Package](https://datapackage.org/standard/data-package/#description)).
```
```suggestion
A description of the schema.
```

peterdesmet commented on content/docs/standard/table-schema.md:
```diff
@@ -266,6 +266,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `created`
+
+The datetime on which the schema was created (cf. [Data Package](https://datapackage.org/standard/data-package/#created)).
```
```suggestion
The datetime on which the schema was created.
```

peterdesmet replied:
I think this could inherit from Data Package. Would add:

> If not specified, the schema inherits from the Data Package if distributed in a Data Package descriptor.

roll replied:
I agree

peterdesmet commented on content/docs/standard/table-schema.md:
```diff
@@ -266,6 +266,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `version`
+
+A version string identifying the version of the schema (cf. [Data Package](https://datapackage.org/standard/data-package/#version)). If not specified, the schema inherits from the Data Package if distributed in a Data Package descriptor.
```
```suggestion
A version string identifying the version of the schema. If not specified, the schema inherits from the Data Package if distributed in a Data Package descriptor.
```

peterdesmet commented on content/docs/standard/table-schema.md:
```diff
@@ -266,6 +266,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `keywords`
+
+An array of string keywords to assist users searching for the schema in catalogs.
```
@pierrecamilleri would you argue this inherits from Data Package or not? If so, please change to:

```suggestion
An array of string keywords to assist users searching for the schema in catalogs. If not specified, the schema inherits from the Data Package if distributed in a Data Package descriptor.
```

peterdesmet commented on content/docs/standard/table-schema.md:
```diff
@@ -266,6 +266,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+ 
+#### `contributors`
+
+The people or organizations that contributed to the schema (cf. [Data Package](https://datapackage.org/standard/data-package/#contributors)). If not specified the schema inherits from the Data Package if distributed in a Data Package descriptor.
```
```suggestion
The people or organizations that contributed to the schema. If not specified the schema inherits from the Data Package if distributed in a Data Package descriptor.
```

peterdesmet commented on profiles/target/2.0/datapackage.json:
```diff
@@ -348,7 +348,7 @@
             "propertyOrder": 70,
             "title": "Home Page",
-            "description": "The home on the web that is related to this data package.",
+            "description": "The home on the web that is related to this resource.",
```
```suggestion
            "description": "The home on the web that is related to this data package.",
```

@roll If these files are rebuild, it should say `descriptor` (general) or `Data Package` (specific) here.

pierrecamilleri replied:
If I remember well this change was due to sharing the definition of `homepage` between data package and schema. Descriptor may be less ambiguous than resource in my opinion, while applying to both.

peterdesmet committed:
Change order cf. Data Package

name, title, description, homepage, version, created, keywords, contributors, examples

peterdesmet commented on content/docs/standard/table-schema.md:
```diff
@@ -266,6 +266,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `version`
+
+A version string identifying the version of the schema (cf. [Data Package](https://datapackage.org/standard/data-package/#version)). If not specified, the schema inherits from the Data Package if distributed in a Data Package descriptor.
```
```suggestion
A version string identifying the version of the schema. If not specified, the schema inherits from the Data Package if distributed in a Data Package descriptor.
```

peterdesmet commented on content/docs/standard/table-schema.md:
```diff
@@ -266,6 +266,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `keywords`
+
+An array of string keywords to assist users searching for the schema in catalogs.
```
Does this inherit from Data Package if not defined? If so, please add:

> If not specified, the schema inherits from the Data Package if distributed in a Data Package descriptor.

peterdesmet commented on profiles/source/dictionary/schema.yaml:
```diff
@@ -81,6 +81,24 @@ tableSchema:
+      "$ref": "#/definitions/version"
+    keywords:
+      "$ref": "#/definitions/keywords"
+    contributors:
```
Four properties from Data Package are not carried over to Schema. I'm fine with waiting for use cases, but wanted to know your opinion on these @pierrecamilleri:

- `id`: potentially useful. In my use case, we opted for `path` instead (the URL where the schema is available), which could be useful to add.
- `licenses`: likely a bit tricky to inherit, would not add for now
- `image`: probably seldom useful
- `sources`: not really applicable here

pierrecamilleri replied:
Sorry I forgot to reply on this one, I agree on every point with you !
We use the `name` property as an ID in our [schema catalog](https://schema.data.gouv.fr/) and we have on our use case no need for an additional field.

peterdesmet reviewed (approved):
Nice work @pierrecamilleri I did the following:

- List the properties in the same order as on Data Package: name, title, description, homepage, version, created, keywords, contributors, examples
- Made some review comments regarding phrasing. I think we can remove the `cf. Data Package` everywhere, but indicate where it inherits from Data Package where applicable.
- Minor update to simplify the CHANGELOG

I've started to use these metadata properties, so I'm looking forward to see this adopted!

pierrecamilleri commented:
Thanks for the quality improvements !

>   I think we can remove the `cf. Data Package` everywhere, 

`Cf. Data Package` was more to avoid repeating stuff in the documentation, for instance the `datapackage.version` gives precision about "It SHOULD conform to the [Semantic Versioning](http://semver.org/) requirements and SHOULD follow the [Data Package Version](https://datapackage.org/recipes/data-package-version) recipe." that we probably don't want to repeat. So I think these links add some clarity.

WDYT ?

> but indicate where it inherits from Data Package where applicable.

Concerning the inheritance, I wonder if we could not rather drop this mention of inheritance, even for `version` and `contributors`. As a user of e.g. `frictionless-r` or `frictionless-py`, this would create the expectation that if I access the `version` or `contributors` property of a schema, it automatically fills a missing field if the schema is part of a package. 

I am not quite sure there is a need for this. The metadata is optional and every creator of datapackage can gauge whether it makes sense to repeat this information, and tooling can choose to look at datapackages properties when relevant.

If you think that dropping it would create confusion, OK for adding it to `keywords` as well.

roll reviewed (approved):
Great work @pierrecamilleri !

pierrecamilleri committed:
Merge remote-tracking branch 'upstream/next' into main

pierrecamilleri committed:
Changelog includes changes for v2.1

pierrecamilleri committed:
Linting

pierrecamilleri commented:
I merged the `next` branch to resolve merge conflicts. In the changelog, I included the changes under the v2.1 header (which release PR is still in Draft status), I hope it is planned to release them under this version number. I corrected the linting errors as well.

pierrecamilleri commented:
Hi all, we added you as a reviewer as you are a [voting member](https://datapackage.org/overview/governance/)  and this PR is tagged "Requires 6 voting approvals" ! We would like this PR to land soon, if you can find the time to make a review it would be great !

PietrH commented on content/docs/standard/table-schema.mdx:
```diff
@@ -202,14 +202,14 @@ In contrast with `field.constraints.unique`, `uniqueKeys` allows to define uniqu
 #### `foreignKeys` {#foreignKeys}
 
-A foreign key is a reference where values in a field (or fields) on the table ('resource' in data package terminology) described by this Table Schema connect to values a field (or fields) on this or a separate table (resource). They are directly modelled on the concept of foreign keys in SQL.
+A foreign key is a reference where values in a field (or fields) on the table ('resource' in Data Package terminology) described by this Table Schema connect to values a field (or fields) on this or a separate table (resource). They are directly modelled on the concept of foreign keys in SQL.
```
```suggestion
A foreign key is a reference where values in a field (or fields) on the table ('resource' in Data Package terminology) described by this Table Schema connect to values a field (or fields) on this or a separate table (resource). This concept is directly modelled on the concept of foreign keys in SQL.
```

```suggestion
A foreign key is a reference where values in a field (or fields) of the table ('resource' in Data Package terminology) described by this Table Schema connect to values a field (or fields) on this or a separate table (resource). They are directly modelled on the concept of foreign keys in SQL.
```

pschumm replied:
```suggestion
A foreign key is a reference where values in a field (or fields) on the table ('resource' in Data Package terminology) described by this Table Schema connect to values a field (or fields) of this or a separate table (resource). They are directly modelled on the concept of foreign keys in SQL.
```

PietrH commented on content/docs/standard/table-schema.mdx:
```diff
@@ -277,6 +277,44 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `homepage`
+
+A URL for the home on the web that is related to the schema.
```
```suggestion
A URL for the web page associated with the schema.
```

PietrH reviewed (commented):
Looks good, I made some small textual suggestions.

khughitt reviewed (approved):
Looks good to me. :+1:

ezwelty reviewed (approved):
End-of-year crunch so admittedly only a cursory review, but looks good to me.

pschumm reviewed (approved):
This looks ok to me.

pierrecamilleri committed:
Apply suggestions from code review

Co-authored-by: Pieter Huybrechts <48065851+PietrH@users.noreply.github.com>

pierrecamilleri committed:
Update table-schema.mdx

Apply review suggestions

pierrecamilleri commented:
Thank you very much all for your quick review ! :pray: 

@roll except for [this](https://github.com/frictionlessdata/datapackage/pull/961/files/a3233f6dfdaea3d6e7a3731f22a36cc5da32c37a#r1767058414) last small wording decision, everything looks ready to me!

roll commented:
Thanks a lot @pierrecamilleri!

I already voted `yes`. Can you please run `npm run generate` having the `next` branch merged and up-to-date so it will applies changes to `v2.1` of the profiles instead of `v2.0`?

pierrecamilleri committed:
npm run generate

pierrecamilleri commented:
`npm run generat`ed!
(sorry for the delay)

johanricher commented:
Great work @pierrecamilleri :clap: Looks like it's ready to merge now :rocket:

roll commented:
Hi @pierrecamilleri,

Sorry for the slow communication due to the holiday season =)

Looks great!

We just need to remove any changes to `public/profiles/2.0.` from the diff to ensure published standard immutability.

pierrecamilleri committed:
Remove changes on v2.0 spec

pierrecamilleri commented:
> We just need to remove any changes to `public/profiles/2.0.` from the diff to ensure published standard immutability.

Done! (Sorry I have missed this)

roll commented:
Amazing! Thanks!

roll commented:
Merged. Here is a live preview of `v2.1` - https://github.com/frictionlessdata/datapackage/pull/1051
